src/tqec/sketchup/zx_graph.py

```diff
@@ -112,18 +127,31 @@ def edges(self) -> list[ZXEdge]:
         """Return a list of edges in the graph."""
         return [data[_EDGE_DATA_KEY] for _, _, data in self._graph.edges(data=True)]
 
+    @property
+    def leaf_nodes(self) -> list[ZXNode]:
+        """Get the leaf nodes of the graph."""
+        return [node for node in self.nodes if self._graph.degree(node.position) == 1]
+
+    @property
+    def isolated_nodes(self) -> list[ZXNode]:
+        """Get the isolated nodes of the graph."""
+        return [node for node in self.nodes if self._graph.degree(node.position) == 0]
+
     def add_node(
         self,
         position: Position3D,
         node_type: NodeType,
+        raise_if_exist: bool = True,
     ) -> None:
         """Add a node to the graph.
 
         Args:
             position: The 3D position of the node.
             node_type: The type of the node.
+            raise_if_exist: Whether to raise an exception if the node already exists.
+
         """
-        if position in self._graph:
+        if raise_if_exist and position in self._graph:
             raise TQECException(f"Node {position} already exists in the graph.")
         self._graph.add_node(position, **{_NODE_DATA_KEY: ZXNode(position, node_type)})
```

Comment:
It was unclear to me what happened if `raise_if_exist == False and position in self._graph` and I had to look it up in the code of `networkx.Graph.add_node`. 
It turns out that, in this case, only the metadata is updated. I think this should be at least written in the documentation, and I would even prefer renaming the `raise_if_exist` flag to something self-explanatory (but I do not have good ideas for a new name, hence my first proposition).